<a href="https://colab.research.google.com/github/madanjha/Machine-Learning/blob/main/Introduction_to_NN_Frameworks_Module3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Module 3: Introduction to NN Frameworks

Problem Statement:
This assignment challenges students to construct and train a neural network to
accurately classify handwritten digits from the MNIST dataset. The goal is to
navigate through the preprocessing of image data, select an appropriate neural
network architecture, and apply optimization techniques to achieve high
classification accuracy

Introduction:
The MNIST dataset is a large database of handwritten digits commonly used for
training various image processing systems. The dataset contains 60,000 training
images and 10,000 testing images, each of which is a 28x28 pixel grayscale
image. The goal is to build a neural network model that can accurately classify
these images into one of the ten digit classes (0 through 9).

Objectives:
Data Preprocessing: Prepare the MNIST dataset for training and testing the
neural network model. This includes normalization of image pixel values and
converting the output labels to a one-hot encoded format.

Model Development: Create a neural network model capable of classifying
28x28 pixel grayscale images of handwritten digits. The model should have an
input layer, one or more hidden layers, and an output layer.

Training: Train the neural network model using the preprocessed training data.
Utilize categorical cross-entropy as the loss function, stochastic gradient descent
as the optimizer, and track accuracy as a performance metric.

Evaluation: Evaluate the performance of the trained model using the testing
dataset, ensuring that the model generalizes well to unseen data.

Prediction: Implement functionality to make predictions on single images or
batches of images, returning the digit class with the highest probability.

Analysis: Analyze the results, understand the model's performance, and explore
ways to improve it if necessary

Challenges and Considerations:

Class Imbalance: Ensure that the dataset does not have a significant class
imbalance that could negatively affect the model’s performance.
If imbalance is present, strategies to mitigate its effects should be considered.

Model Complexity: Choose an appropriate model complexity that is capable of
learning from the data without overfitting. This may involve experimenting with
different numbers of layers and units, as well as regularization techniques.

Activation Functions: Use appropriate activation functions in the hidden layers
(ReLU is commonly used for image data) and softmax activation in the output
layer to convert logits to probabilities.

Evaluation Metrics: Focus on accuracy as the primary evaluation metric, but
also be open to exploring other metrics that could provide additional insights into
the model’s performance.

Hyperparameter Tuning: Experiment with different hyperparameters such as
learning rate, batch size, and number of epochs to optimize the model’s
performance.

Predictions and Interpretability: Provide clear and interpretable predictions,
ensuring that the output probabilities are well-calibrated and meaningful.

# Task
Classify handwritten digits from the MNIST dataset by loading and preprocessing the data, building and training a neural network, and evaluating its performance.

## Load and Preprocess Data

### Subtask:
Load the MNIST dataset, normalize image pixel values to a [0, 1] range, and convert labels to one-hot encoded format. Ensure data is reshaped appropriately for the neural network input.


**Reasoning**:
To begin the data preprocessing, I will load the MNIST dataset, normalize the image pixel values, reshape the images for network input, and convert the labels to one-hot encoding, all within a single code block for efficiency.



In [1]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical

# 1. Load the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()

print(f"Original training images shape: {x_train.shape}")
print(f"Original testing images shape: {x_test.shape}")
print(f"Original training labels shape: {y_train.shape}")
print(f"Original testing labels shape: {y_test.shape}")

# 2. Normalize pixel values
x_train = x_train.astype('float32') / 255.0
x_test = x_test.astype('float32') / 255.0

# 3. Reshape the image data
# For a simple dense network, images are flattened from 28x28 to 784
num_pixels = x_train.shape[1] * x_train.shape[2]
x_train = x_train.reshape(x_train.shape[0], num_pixels)
x_test = x_test.reshape(x_test.shape[0], num_pixels)

print(f"Reshaped training images shape: {x_train.shape}")
print(f"Reshaped testing images shape: {x_test.shape}")

# 4. Convert labels to one-hot encoded format
num_classes = 10
y_train = to_categorical(y_train, num_classes)
y_test = to_categorical(y_test, num_classes)

print(f"One-hot encoded training labels shape: {y_train.shape}")
print(f"One-hot encoded testing labels shape: {y_test.shape}")

print("MNIST dataset loaded, normalized, reshaped, and labels one-hot encoded successfully.")

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 2s 0us/step
Original training images shape: (60000, 28, 28)
Original testing images shape: (10000, 28, 28)
Original training labels shape: (60000,)
Original testing labels shape: (10000,)
Reshaped training images shape: (60000, 784)
Reshaped testing images shape: (10000, 784)
One-hot encoded training labels shape: (60000, 10)
One-hot encoded testing labels shape: (10000, 10)
MNIST dataset loaded, normalized, reshaped, and labels one-hot encoded successfully.


## Define Neural Network Model

### Subtask:
Construct a neural network model with an input layer suitable for 28x28 grayscale images, one or more hidden layers with ReLU activation, and an output layer with 10 units and softmax activation for classification.


**Reasoning**:
To construct the neural network model as specified, I will import the necessary Keras components, define a Sequential model, and add the input, hidden, and output Dense layers with the correct activation functions and unit counts.



In [2]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# 1. Instantiate a Sequential model
model = Sequential()

# 2. Add a Dense layer as the input layer
# num_pixels was defined in the previous step as 784
model.add(Dense(128, activation='relu', input_shape=(num_pixels,)))

# 3. Add at least one more Dense hidden layer
model.add(Dense(64, activation='relu'))

# 4. Add a final Dense output layer
# num_classes was defined in the previous step as 10
model.add(Dense(num_classes, activation='softmax'))

# Print the model summary to confirm its structure
print("Neural Network Model Summary:")
model.summary()

print("Neural network model constructed successfully.")

Neural Network Model Summary:


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 10)             │           650 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 109,386 (427.29 KB)

 Trainable params: 109,386 (427.29 KB)

 Non-trainable params: 0 (0.00 B)

Neural network model constructed successfully.


**Reasoning**:
The previous output included a Keras UserWarning regarding the `input_shape` argument in a `Dense` layer within a `Sequential` model. To address this, I will explicitly add an `Input` layer as the first layer of the model.



In [3]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Input

# 1. Instantiate a Sequential model
model = Sequential()

# Add an Input layer explicitly to resolve the UserWarning
model.add(Input(shape=(num_pixels,)))

# 2. Add a Dense layer as the input layer (now implicitly connected to Input layer)
model.add(Dense(128, activation='relu'))

# 3. Add at least one more Dense hidden layer
model.add(Dense(64, activation='relu'))

# 4. Add a final Dense output layer
# num_classes was defined in the previous step as 10
model.add(Dense(num_classes, activation='softmax'))

# Print the model summary to confirm its structure
print("Neural Network Model Summary:")
model.summary()

print("Neural network model constructed successfully.")

Neural Network Model Summary:


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_3 (Dense)                 │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 10)             │           650 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 109,386 (427.29 KB)

 Trainable params: 109,386 (427.29 KB)

 Non-trainable params: 0 (0.00 B)

Neural network model constructed successfully.


## Compile Neural Network Model

### Subtask:
Compile the neural network model using `categorical_crossentropy` as the loss function, `Adam` optimizer, and `accuracy` as the metric.

**Reasoning**:
Following the subtask instructions, I will compile the neural network model using the specified loss function, optimizer, and metrics.



In [4]:
from tensorflow.keras.optimizers import Adam

# Compile the model
model.compile(
    loss='categorical_crossentropy',
    optimizer=Adam(), # Using Adam optimizer as it's a common and effective choice
    metrics=['accuracy']
)

print("Neural network model compiled successfully with Adam optimizer, categorical_crossentropy loss, and accuracy metric.")

Neural network model compiled successfully with Adam optimizer, categorical_crossentropy loss, and accuracy metric.


## Train Neural Network Model

### Subtask:
Train the neural network model using the preprocessed training data. Utilize appropriate hyperparameters like epochs and batch size, and ensure accuracy is tracked as a performance metric during training.

**Reasoning**:
Now that the model has been defined and compiled, I will train the neural network using the `model.fit()` method. I will use the preprocessed `x_train` and `y_train` for training, and `x_test` and `y_test` for validation to monitor performance and potential overfitting, along with setting `epochs` and `batch_size`.



In [5]:
print("Training the neural network model...")

# Define training hyperparameters
epochs = 10
batch_size = 128

# Train the model
history = model.fit(
    x_train,
    y_train,
    epochs=epochs,
    batch_size=batch_size,
    validation_data=(x_test, y_test),
    verbose=1
)

print(f"Neural network model trained for {epochs} epochs with a batch size of {batch_size}.")

Training the neural network model...
Epoch 1/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 4s 6ms/step - accuracy: 0.8344 - loss: 0.5920 - val_accuracy: 0.9519 - val_loss: 0.1590
Epoch 2/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9559 - loss: 0.1504 - val_accuracy: 0.9662 - val_loss: 0.1132
Epoch 3/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 4s 8ms/step - accuracy: 0.9707 - loss: 0.1016 - val_accuracy: 0.9700 - val_loss: 0.0991
Epoch 4/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 3s 7ms/step - accuracy: 0.9780 - loss: 0.0741 - val_accuracy: 0.9720 - val_loss: 0.0917
Epoch 5/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9828 - loss: 0.0588 - val_accuracy: 0.9725 - val_loss: 0.0825
Epoch 6/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9868 - loss: 0.0451 - val_accuracy: 0.9761 - val_loss: 0.0769
Epoch 7/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 7s 9ms/step - accuracy: 0.9894 - loss: 0.0346 - val_accuracy: 0.9770 - val_loss: 0.0738
Epoch 8/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9

## Evaluate Neural Network Model

### Subtask:
Evaluate the performance of the trained model using the testing dataset, ensuring that the model generalizes well to unseen data. This includes reporting the loss and accuracy on the test set.

**Reasoning**:
To evaluate the trained model on the unseen test data, I will use the `model.evaluate()` method with the `x_test` and `y_test` datasets.



In [6]:
print("Evaluating the neural network model on the test data...")

loss, accuracy = model.evaluate(x_test, y_test, verbose=0)

print(f"Test Loss: {loss:.4f}")
print(f"Test Accuracy: {accuracy:.4f}")
print("Model evaluation complete.")

Evaluating the neural network model on the test data...
Test Loss: 0.0925
Test Accuracy: 0.9743
Model evaluation complete.


## Implement Prediction Functionality

### Subtask:
Develop a function to make predictions on new, unseen single images or batches of images, returning the predicted digit class with the highest probability.


**Reasoning**:
To implement the prediction functionality, I will first select a few sample images from the `x_test` dataset and their corresponding true labels. Then, I will ensure these samples are reshaped to match the model's expected input format (flattened 784 pixels). After that, I will use the trained model to make predictions and interpret the probability distributions to determine the predicted digit class for each sample. Finally, I will print both the predicted and true digits for each sample to evaluate the model's performance on individual examples.

**Reasoning**:
Following the reasoning, I will now implement the prediction functionality by selecting sample images, reshaping them if necessary, making predictions with the trained model, and then printing the true and predicted labels.



In [7]:
import numpy as np

print("Implementing prediction functionality...")

# 1. Select a few sample images from x_test
# Let's pick the first 5 images for demonstration
sample_indices = range(5)
sample_images = x_test[sample_indices]
sample_true_labels = y_test[sample_indices]

# 2. Reshape these sample images to be compatible with the model's input
# (Already flattened during preprocessing, so this step is for clarity/reassurance)
# Model expects (num_samples, num_pixels)
# x_test is already (10000, 784), so no reshaping needed here for these specific samples.
# If we were to take raw 28x28 images, we'd do:
# sample_images_reshaped = sample_images.reshape(sample_images.shape[0], num_pixels)

# 3. Use the trained model to make predictions on these sample images
predictions = model.predict(sample_images)

# 4. Interpret the model's output to determine the predicted digit class
predicted_classes = np.argmax(predictions, axis=1)
true_classes = np.argmax(sample_true_labels, axis=1)

# 5. Print the predicted digit along with the true digit for each sample
print("\nPredictions for sample images:")
for i in range(len(sample_indices)):
    print(f"Sample {i+1}: True Digit: {true_classes[i]}, Predicted Digit: {predicted_classes[i]}, Probabilities: {predictions[i]}")

print("Prediction functionality implemented and tested on sample images.")

Implementing prediction functionality...
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 200ms/step

Predictions for sample images:
Sample 1: True Digit: 7, Predicted Digit: 7, Probabilities: [3.1656992e-08 3.1931648e-11 7.0522582e-08 1.9745321e-05 4.6721363e-11
 6.3786025e-11 9.0239131e-16 9.9997699e-01 7.2846184e-08 3.0829904e-06]
Sample 2: True Digit: 2, Predicted Digit: 2, Probabilities: [4.8793108e-10 2.3248920e-05 9.9997568e-01 1.0363420e-06 7.6558821e-17
 9.5380792e-10 1.9495412e-10 1.5224546e-13 1.0857959e-09 1.4300997e-16]
Sample 3: True Digit: 1, Predicted Digit: 1, Probabilities: [6.7451560e-07 9.9939978e-01 3.7550149e-04 2.1080655e-06 3.7997779e-05
 9.6176052e-07 1.3390533e-06 1.3248774e-04 4.8803049e-05 2.4501270e-07]
Sample 4: True Digit: 0, Predicted Digit: 0, Probabilities: [9.9999869e-01 1.2214642e-09 3.3147694e-07 1.6420154e-08 3.5174088e-09
 3.2282326e-11 7.3473370e-07 2.7698047e-07 1.0206711e-12 1.2539053e-09]
Sample 5: True Digit: 4, Predicted Digit: 4, Probabilities: [1.6604445e-06 1

## Analyze and Improve Model

### Subtask:
Analyze the model's performance, identify areas for improvement (e.g., through hyperparameter tuning, model complexity adjustments, or regularization), and iterate on the model design if needed.


## Summary:

### Data Analysis Key Findings
*   The MNIST dataset was successfully loaded and preprocessed. Image pixel values were normalized to a \[0, 1] range, and 28x28 images were reshaped into 1D arrays of 784 pixels. Labels were converted to a 10-class one-hot encoded format.
*   A neural network model was constructed with an input layer, two dense hidden layers (128 and 64 units, both with ReLU activation), and an output layer (10 units with softmax activation). The model comprised 109,386 trainable parameters.
*   The model was compiled using `categorical_crossentropy` as the loss function, the `Adam` optimizer, and `accuracy` as the metric.
*   After training for 10 epochs with a batch size of 128, the model achieved a final training accuracy of 0.9936 and a validation accuracy of 0.9743.
*   The model was evaluated on the test dataset, achieving a Test Loss of 0.0925 and a Test Accuracy of 0.9743.
*   Prediction functionality was successfully implemented and demonstrated on 5 sample images from the test set, where the model correctly predicted all 5 digits.

### Insights or Next Steps
*   The model exhibits strong performance, achieving a test accuracy of approximately 97.43%. This suggests it generalizes well to unseen data.
*   Consider exploring more complex model architectures, such as Convolutional Neural Networks (CNNs), which are typically better suited for image classification tasks and may yield even higher accuracy.


**Evaluae the trained model's Performance**

In [8]:
print(f"Test Loss: {loss:.4f}")
print(f"Test Accuracy: {accuracy:.4f}")

Test Loss: 0.0925
Test Accuracy: 0.9743
